lib/icu.js

```diff
@@ -0,0 +1,209 @@
+    }
+  },
+  'latin1': {
+    'ascii': (source) => {
+      return icu.convert('us-ascii', 'iso8859-1', source);
+    },
+    'utf8': (source) => {
+      return icu.convert('utf-8', 'iso8859-1', source);
+    },
+    'utf16le': (source) => {
+      return icu.convertToUcs2('iso8859-1', source);
+    }
+  },
+  'utf8': {
+    'ascii': (source) => {
+      return icu.convert('us-ascii', 'utf-8', source);
+    },
+    'latin1': (source) => {
+      return icu.convert('iso-8859-1', 'utf-8', source);
+    },
+    'utf16le': icu.convertToUcs2FromUtf8,
+  },
+  'utf16le': {
+    'ascii': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertFromUcs2('us-ascii', source);
+    },
+    'latin1': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertFromUcs2('iso-8859-1', source);
+    },
+    'utf8': (source) => {
+      if (source.length % 2 !== 0)
+        throw new TypeError('Invalid UCS2 Buffer');
+      return icu.convertToUtf8FromUcs2(source);
+    }
+  }
+};
+
+// Transcodes the Buffer from one encoding to another, returning a new
+// Buffer instance.
+exports.transcode = function transcode(source, from_enc, to_enc) {
+  if (!source || !(source.buffer instanceof ArrayBuffer))
+    throw new TypeError('"source" argument must be a Buffer');
+  if (source.length === 0) return Buffer.alloc(0);
+
+  from_enc = normalizeEncoding(from_enc);
+  to_enc = normalizeEncoding(to_enc);
+
+  if (from_enc === to_enc)
+    return Buffer.from(source);
+
+  const cnv_from = conversions[from_enc];
+
+  if (cnv_from) {
+    const cnv_to = cnv_from[to_enc];
+    if (cnv_to)
+      return cnv_to(source);
+  }
+  throw new TypeError(`Unsupported conversion: ${from_enc} to ${to_enc}`);
+};
+
+// Returns the unicode codepoint at a specific offset location in the buffer,
+// based on the given encoding
+exports.codePointAt = function(buf, pos, encoding) {
+  if (!buf || !(buf.buffer instanceof ArrayBuffer))
+    throw new TypeError('First argument must be a Buffer');
+  if (pos < 0 || pos >= buf.length)
+    return;
+  pos >>>= 0;
+
+  encoding = normalizeEncoding(encoding || 'utf8');
+
+  if (encoding === 'binary' || encoding === 'ascii')
+    return buf[pos];
+  if (encoding !== 'utf8' && encoding !== 'ucs2')
+    throw new TypeError(`Unsupported Encoding: ${encoding}`);
+
+  return icu.getCodePointAt(buf, encoding === 'utf8', pos);
+};
+
+// Returns the char at a specific offset location in the buffer, based
+// on the given encoding.
+exports.charAt = function bufferCharAt(buf, pos, encoding) {
+  if (!buf || !(buf.buffer instanceof ArrayBuffer))
+    throw new TypeError('First argument must be a Buffer');
+  if (pos < 0 || pos >= buf.length)
+    return;
+
+  pos >>>= 0;
+  encoding = normalizeEncoding(encoding || 'utf8');
+
+  if (encoding === 'binary' || encoding === 'ascii')
+    return String.fromCharCode(buf[pos]);
```

Comment:
Should we do any coercion if encoding is 'ascii' but the code point lies outside the correct range?